test/integration/http2_integration_test.cc

```diff
@@ -1598,61 +1623,54 @@ void Http2FloodMitigationTest::beginSession() {
   startHttp2Session();
 }
 
 // Verify that the server detects the flood of the given frame.
-void Http2FloodMitigationTest::floodServer(const Http2Frame& frame, const std::string& flood_stat) {
-  // pack the as many frames as we can into 16k buffer
-  const int FrameCount = (16 * 1024) / frame.size();
-  std::vector<char> buf(FrameCount * frame.size());
+void Http2FloodMitigationTest::floodServer(const Http2Frame& frame, const std::string& flood_stat,
+                                           uint32_t num_frames) {
+  // make sure all frames can fit into 16k buffer
+  ASSERT_LE(num_frames, (16u * 1024u) / frame.size());
+  std::vector<char> buf(num_frames * frame.size());
   for (auto pos = buf.begin(); pos != buf.end();) {
     pos = std::copy(frame.begin(), frame.end(), pos);
   }
 
-  tcp_client_->readDisable(true);
-  int64_t total_bytes_sent = 0;
-  // If the flood protection is not working this loop will keep going
-  // forever until it is killed by blaze timer or run out of memory.
-  // Add early stop if we have sent more than 100M of frames, as it this
-  // point it is obvious something is wrong.
-  while (total_bytes_sent < TransmitThreshold && tcp_client_->connected()) {
-    ASSERT_TRUE(tcp_client_->write({buf.begin(), buf.end()}, false, false));
-    total_bytes_sent += buf.size();
-  }
+  ASSERT_TRUE(tcp_client_->write({buf.begin(), buf.end()}, false, false));
+
+  // Envoy's flood mitigation should kill the connection
+  tcp_client_->waitForDisconnect();
 
-  EXPECT_LE(total_bytes_sent, TransmitThreshold) << "Flood mitigation is broken.";
   EXPECT_EQ(1, test_server_->counter(flood_stat)->value());
   test_server_->waitForCounterGe("http.config_test.downstream_cx_delayed_close_timeout", 1);
 }
 
 // Verify that the server detects the flood using specified request parameters.
 void Http2FloodMitigationTest::floodServer(absl::string_view host, absl::string_view path,
                                            Http2Frame::ResponseStatus expected_http_status,
-                                           const std::string& flood_stat) {
+                                           const std::string& flood_stat, uint32_t num_frames) {
   uint32_t request_idx = 0;
   auto request = Http2Frame::makeRequest(request_idx, host, path);
   sendFrame(request);
   auto frame = readFrame();
   EXPECT_EQ(Http2Frame::Type::Headers, frame.type());
   EXPECT_EQ(expected_http_status, frame.responseStatus());
-  tcp_client_->readDisable(true);
-  uint64_t total_bytes_sent = 0;
-  while (total_bytes_sent < TransmitThreshold && tcp_client_->connected()) {
+  writev_matcher_->setWritevReturnsEgain();
+  for (uint32_t frame = 0; frame < num_frames; ++frame) {
     request = Http2Frame::makeRequest(++request_idx, host, path);
     sendFrame(request);
-    total_bytes_sent += request.size();
   }
-  EXPECT_LE(total_bytes_sent, TransmitThreshold) << "Flood mitigation is broken.";
+  tcp_client_->waitForDisconnect();
```

Comment:
I think that in cases where flood_stat is not set there may be potential for the close condition being something other than flooding.  I don't know if checking for http.config_test.downstream_cx_delayed_close_timeout mitigates that hazard.

Reply:
Generally true. However it is not set in one test only `TooManyStreams` which is not strictly a flood test. I'll think about refactoring this such that stat name is always specified.